src/membership.rs

```diff
@@ -0,0 +1,368 @@
+        }
+
+        /// Change member's handle.
+        fn change_member_handle(origin, handle: Vec<u8>) {
+            let who = ensure_signed(origin)?;
+            Self::_change_member_handle(&who, handle)?;
+        }
+
+        fn batch_change_member_profile(origin, user_info: UserInfo) {
+            let who = ensure_signed(origin)?;
+            user_info.avatar_uri.map(|uri| Self::_change_member_avatar(&who, &uri)).ok_or("uri not changed");
+            user_info.about.map(|about| Self::_change_member_about_text(&who, &about)).ok_or("about text not changed");
+            user_info.handle.map(|handle| Self::_change_member_handle(&who, handle)).ok_or("handle not changed");
+        }
+
+        /// Buy the default membership (if it is active) and only provide handle - for testing
+        fn buy_default_membership_testing(origin, handle: Vec<u8>) {
+            Self::buy_membership(origin, T::PaidTermId::sa(0), UserInfo {
+                handle: Some(handle.clone()),
+                avatar_uri: None,
+                about: None
+            })?;
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    fn ensure_not_member(who: &T::AccountId) -> dispatch::Result {
+        ensure!(!<MemberIdByAccountId<T>>::exists(who), "Account already associated with a membership");
+        Ok(())
+    }
+
+    fn ensure_is_member(who: &T::AccountId) -> Result<T::MemberId, &'static str> {
+        let member_id = Self::member_id_by_account_id(who).ok_or("no member id found for accountid")?;
+        Ok(member_id)
+    }
+
+    fn ensure_has_profile(who: &T::AccountId) -> Result<Profile<T>, &'static str> {
+        let member_id = Self::ensure_is_member(who)?;
+        let profile = Self::member_profile(&member_id).ok_or("member profile not found")?;
+        Ok(profile)
+    }
+
+    fn ensure_active_terms_id(terms_id: T::PaidTermId) -> Result<PaidMembershipTerms<T>, &'static str> {
+        let active_terms = Self::active_paid_membership_terms();
+        ensure!(active_terms.iter().any(|&id| id == terms_id), "paid terms id not active");
+        let terms = Self::paid_membership_terms_by_id(terms_id).ok_or("paid membership term id does not exist")?;
+        Ok(terms)
+    }
+
+    fn ensure_unique_handle(handle: &Vec<u8> ) -> dispatch::Result {
+        ensure!(!<Handles<T>>::exists(handle), "handle already registered");
+        Ok(())
+    }
+
+    fn validate_handle(handle: &Vec<u8>) -> dispatch::Result {
+        ensure!(handle.len() >= Self::min_handle_length() as usize, "handle too short");
+        ensure!(handle.len() <= Self::max_handle_length() as usize, "handle too long");
+        Ok(())
+    }
+
+    fn validate_text(text: &Vec<u8>) -> Vec<u8> {
+        let mut text = text.clone();
+        text.truncate(Self::max_about_text_length() as usize);
+        text
+    }
+
+    fn validate_avatar(uri: &Vec<u8>) -> Vec<u8> {
+        let mut uri = uri.clone();
+        uri.truncate(Self::max_avatar_uri_length() as usize);
```

Comment:
Not sure that truncate is a good approach for avatar: if you truncate an URL then it became invalid.

Reply:
We have to limit the length somehow. We can accommodate a longer url  upto 2000 ? https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers

Reply:
The same as with a handle:
`ensure!(uri.len() <= Self::max_avatar_uri_length() as usize, "avatar uri is too long");`